Major improvement in how and when the dashboard is drawn

yubiuser opened the pull request:
**What does this PR aim to accomplish?:**

Major improvement in how and when the dashboard is drawn.
1) Reacts to terminal resizes and immediately redraws the dashboard
2) allows to set `xoff`and `yoff`as offsets to move the dashboard **within** the terminal. Fixes https://github.com/pi-hole/PADD/issues/266

**How does this PR accomplish the above?:**

1) traps `WINCH` and executes `TerminalResize()`
2) uses ANSI escape sequences instead of `tput` (which has a lot of overhead), making the script more portable
 3) further separates collecting information and displaying them (added `GenerateSizeDependendOutput`)
4) adds a `CleanExit` function to restore the terminal functions on exit

P.S. Adding `y-offset` was easy, as it was simply moving the "cursor" y-times down. I tried implementing `x-offset` with `left margin` functions (`DECLRMM`, `DECSLRM` and `DECOM`). Unfortunately, those are not supported by most terminal emulators (got it working on `xterm`), see https://terminalguide.namepad.de/mode/ So I needed to implement it on every line of the dasboard individually.


---
**By submitting this pull request, I confirm the following:** 

1. I have read and understood the [contributors guide](https://docs.pi-hole.net/guides/github/contributing/), as well as this entire template. I understand which branch to base my commits and Pull Requests against. 
2. I have commented my proposed changes within the code and I have tested my changes.
3. I am willing to help maintain this change if there are issues with it later.
4. It is compatible with the [EUPL 1.2 license](https://opensource.org/licenses/EUPL-1.1)
5. I have squashed any insignificant commits. ([`git rebase`](http://gitready.com/advanced/2009/02/10/squashing-commits-with-rebase.html))

---
- [x] I have read the above and my PR is ready for review. _Check this box to confirm_

dschaper commented:
Nice!

yubiuser commented:
x/y Offset

![Screenshot at 2022-10-24 09-54-12](https://user-images.githubusercontent.com/26622301/197475940-92332838-1e19-4cce-ac99-396def488809.png)
____

Resizing
(There is still a bit flickering, esp. with the items generated by `GenerateSizeDependendOutput()`)
![Peek 2022-10-24 09-55](https://user-images.githubusercontent.com/26622301/197475996-dad994db-7882-4ea4-a6f7-93cda1c8228d.gif)

yubiuser commented:
Got rid of the flickering.
![Peek 2022-10-24 21-41](https://user-images.githubusercontent.com/26622301/197611652-8937ba96-4d75-4245-a155-ab10516a6954.gif)

Two modifications were necessary:
1) `TerminalResize` does not re-draw the dashboard itself, but instead kills the `sleep` function in `NormalPADD()`. This will then re-draw the dashboard right away.
2) Added a small delay before checking the terminal size. This gives the kernel a bit time to report the new size and further reduced "flickering"

DL6ER commented:
I tested it and it seems to work as expected. How are x and y offset supposed to work? Are they quantities a user has to supply or is it rather meant to automagically center the most suitable PADD variant on any given terminal (may it be as large as it wants).

**edit** NVM, I found it is a manual process:
``` bash
setOffsets(){
    # sets x/y-offsets (for shifting the Dashboard within the screen) if CLI arguments were passed
    i=1;
    j=$#;
    while [ $i -le $j ]
    do
        case "$1" in
            "-xoff"  ) xOffset="$2";;
            "-yoff"  ) yOffset="$2";;
        esac
        i=$((i + 1));
        shift 1;
    done
}
```

but how about the idea of the automatic centering? I could even imagine that this would be a nice by-default option with the possibility to specify `./padd.sh -no-center`

yubiuser commented:
@rdwebdesign has a follow up PR based on this one that will automatically center.

rdwebdesign commented:
> **rdwebdesign** has a follow up PR based on this one that will automatically center.

I'm just waiting this PR to be approved.

DL6ER reviewed (changes requested):
Okay, after some further testing, I see another issue when defining offsets. When I specify an Y-offset that is would cause a reduction to the next smaller screen size, the offset is not (or only partially) acknowledged. I guess that's by design. (?)

However, the same does not work for X-offset. Given my terminal is wide enough for a given size, this size will be picked even if I have an X-offset resulting in wrapping to the next line and destroying the entire display.

### Pictures

Initial terminal size is 83x30
![Screenshot from 2022-11-19 18-48-43](https://user-images.githubusercontent.com/16748619/202864779-7eee1657-645f-4d46-accb-22cc8a0b7c0b.png)
Reduction in Y by 1 (83x29) removes the empty line at the bottom (okay):
![Screenshot from 2022-11-19 18-49-25](https://user-images.githubusercontent.com/16748619/202864783-3f93a3f8-129a-4ec5-9100-6b2068d838ff.png)
Further reductions in Y (83x27 in this example) reduces the top offset (okay):
![Screenshot from 2022-11-19 18-50-06](https://user-images.githubusercontent.com/16748619/202864795-08d9b4f2-c72c-428b-9c48-49479a17ff7b.png)

---

(back to initial size of 83x30)
![Screenshot from 2022-11-19 18-48-43](https://user-images.githubusercontent.com/16748619/202864800-a6bc815c-b004-4ad6-adeb-025acef03756.png)
Reduction in X by 1 (82x30) causes two additional linebreaks in FTL and SYSTEM causing the Y-offset to decrease:
![Screenshot from 2022-11-19 18-51-26](https://user-images.githubusercontent.com/16748619/202864829-bcd112dc-257b-494a-a58c-fa52b6973098.png)
Further reduction in X (81x30) finally causes destruction all over the place:
![Screenshot from 2022-11-19 18-52-02](https://user-images.githubusercontent.com/16748619/202864835-eba443b6-0932-49c8-8db3-26f453660109.png)

rdwebdesign commented:
I can confirm with a different approach:
starting PADD with a offset bigger than the "margins" will cause problems.

Example:
mega is 80x26.

With a window of 85x30, there are 5 columns more than "mega" needs. If I start PADD with `-xoff 6` the lines will break.

The same occurs with `-yoff 6`, but in this case, the "break" is just rolling the lines up.
The final offset will be smaller, but no visual break.

rdwebdesign commented:
I think I got it.

At the end of `SizeChecker()` function, add something like this:

```
    # Limit the offset to avoid breaks
    xMaxOffset=$(($console_width - $width))
    yMaxOffset=$(($console_height - $height))
    if [ "$xOffset" -gt "$xMaxOffset" ]; then
        xOffset="$xMaxOffset"
    fi
    if [ "$yOffset" -gt "$yMaxOffset" ]; then
        yOffset="$yMaxOffset"
    fi
```

DL6ER commented:
Temporarily converting to a draft until we've found a solution. @rdwebdesign's suggestion doesn't solve all cases.

rdwebdesign commented:
My solution only works for my new code (including the center function and all changes I made there).

rdwebdesign commented:
Try this code, please:

```sh
SizeChecker(){
    # adding a tiny delay here to to give the kernel a bit time to
    # report new sizes correctly after a terminal resize
    # this reduces "flickering" of GenerateSizeDependendOutput() items
    # after a terminal re-size
    sleep 0.1
    console_height=$(stty size | awk '{ print $1 }')
    console_width=$(stty size | awk '{ print $2 }')

    # Below Pico. Gives you nothing...
    if [ "$console_width" -lt "20" ] || [ "$console_height" -lt "10" ]; then
        # Nothing is this small, sorry
        printf "%b" "${check_box_bad} Error!\n    PADD isn't\n    for ants!\n"
        exit 1
    # Below Nano. Gives you Pico.
    elif [ "$console_width" -lt "24" ] || [ "$console_height" -lt "12" ]; then
        padd_size="pico"
        width=20
        height=10
    # Below Micro, Gives you Nano.
    elif [ "$console_width" -lt "30" ] || [ "$console_height" -lt "16" ]; then
        padd_size="nano"
        width=24
        height=12
    # Below Mini. Gives you Micro.
    elif [ "$console_width" -lt "40" ] || [ "$console_height" -lt "18" ]; then
        padd_size="micro"
        width=30
        height=16
    # Below Tiny. Gives you Mini.
    elif [ "$console_width" -lt "53" ] || [ "$console_height" -lt "20" ]; then
        padd_size="mini"
        width=40
        height=18
    # Below Slim. Gives you Tiny.
    elif [ "$console_width" -lt "60" ] || [ "$console_height" -lt "21" ]; then
        padd_size="tiny"
        width=53
        height=20
    # Below Regular. Gives you Slim.
    elif [ "$console_width" -lt "80" ] || [ "$console_height" -lt "22" ]; then
        padd_size="slim"
        width=60
        height=21
    # Below Mega. Gives you Regular.
    elif [ "$console_width" -lt "80" ] || [ "$console_height" -lt "26" ]; then
        padd_size="regular"
        width=60
        height=22
    # Mega
    else
        padd_size="mega"
        width=80
        height=26
    fi

    # Limit the offset to avoid breaks
    xMaxOffset=$(($console_width - $width))
    yMaxOffset=$(($console_height - $height))

    if [ "$xOffset" -gt "$xMaxOffset" ]; then
        xOffset="$xMaxOffset"
    fi

    if [ "$yOffset" -gt "$yMaxOffset" ]; then
        yOffset="$yMaxOffset"
    fi
}
```

DL6ER commented:
@rdwebdesign Thanks, that surely improves the situation but it overwrites the offsets. Once the terminal got small enough, it won't add the offset if the terminal is made larger again. Surely this isn't an issue if we're dealing with displays with constant resolution, however, it feels strange when we have interactive terminals. This fix will be simple and I'll apply and push it ([47d28c2](https://github.com/pi-hole/PADD/pull/268/commits/47d28c2a6bc671abf5beb566eebbed69218eba39)).

However, there is another issue. Assume no offsets are set (to make it easier):

- 80x26 shows `mega` (80x26 sized)
- 80x25 shows `regular` (60x22 sized)
- 79x25 shows `slim` (60x21 sized) although `regular` would easily fit here as well

My last commit ([524ec8f](https://github.com/pi-hole/PADD/pull/268/commits/524ec8fa2f3d9fd12633840b16a88aa276a6f423)) fixes this.

DL6ER commented:
Sorry that this comes piece-by-piece but there is just so much to test on this... I found a typo that broke `tiny` (fixed in [2735a54](https://github.com/pi-hole/PADD/pull/268/commits/2735a544d24aef1ac42a40aa1044691c18b786d0)) but then wondered why `small` actually shows less info than `tiny`:

![Screenshot from 2022-11-20 09-26-37](https://user-images.githubusercontent.com/16748619/202892757-4c8e3c68-cac3-452d-85d6-15bba8cf75f2.png)
![Screenshot from 2022-11-20 09-26-41](https://user-images.githubusercontent.com/16748619/202892762-eb654201-b131-4853-84e4-760a8a167d30.png)

There would be enough space in `slim` to display the exact same version information (`slim` is both taller and wider than `tiny`). I'm not changing this right now as it is not a bug but a style decision.

rdwebdesign commented on padd.sh:
```diff
@@ -959,7 +959,7 @@ SizeChecker(){
         height=20
     # Below Regular. Gives you Slim.
-    elif [ "$console_width" -lt "80" ] || [ "$console_height" -lt "22" ]; then
+    elif [ "$console_width" -lt "60" ] || [ "$console_height" -lt "22" ]; then
```
You are setting `width=60` if `$console_width` is **less than** 60.

If the console width are 58, this will be true and the screen will break because PADD will be 60 columns wide.

DL6ER replied:
No. The code can actually never get here as the `[ "$console_width" -lt "60" ]` will fire in the `elif` immediately above. I just left it there as this matches the overall style of this function.

DL6ER commented on padd.sh:
```diff
@@ -900,40 +918,82 @@ BarGenerator() {
+        width=80
+        height=26
+    # Below Mega. Gives you Regular.
+    elif [ "$console_width" -ge "60" ] && [ "$console_height" -ge "26" ]; then
```
```suggestion
    elif [ "$console_width" -ge "60" ] && [ "$console_height" -ge "22" ]; then
```

rdwebdesign commented on padd.sh:
```diff
@@ -1128,36 +1212,89 @@ DisplayHelp() {
-  # (From Pull request #8 https://github.com/jpmck/PADD/pull/8)
-  setterm -cursor off
-  trap '{ setterm -cursor on ; echo "" ; exit 0 ; }' INT TERM EXIT
+setOffsets(){
+    # sets x/y-offsets (for shifting the Dashboard within the screen) if CLI arguments were passed
+    i=1;
+    j=$#;
+    while [ $i -le $j ]
+    do
+        case "$1" in
+            "-xoff"  ) xOffset=$2; xOffOrig=$2;;
+            "-yoff"  ) yOffset=$2; yOffOrig=$2;;
+        esac
+        i=$((i + 1));
+        shift 1;
+    done
+}
```
You don't need to use these 2 auxiliary variables. 
`shift` will automatically reduce the value of `$#` until there is no more parameters.

```suggestion
setOffsets(){
    # sets x/y-offsets (for shifting the Dashboard within the screen) if CLI arguments were passed
    while [ $# -gt 0 ]
    do
        case "$1" in
            "-xoff"  ) xOffset="$2";;
            "-yoff"  ) yOffset="$2";;
        esac
        shift 2;
    done
}
```

DL6ER replied:
@rdwebdesign We still need `xOffOrig=$2` and `yOffOrig=$2` to restore offsets after terminals have been made narrower and then wider again. See https://github.com/pi-hole/PADD/pull/268/commits/47d28c2a6bc671abf5beb566eebbed69218eba39

rdwebdesign replied:
I edited my original post... the variables are here.

yubiuser commented:
Thanks for your input. Works really nice now. I added rd's suggestion about removing the auxiliary variables and rebased on `development`.

Ready for re-review

rdwebdesign commented:
I approve it too soon...

The `mini` size is getting broken:
![padd_mini](https://user-images.githubusercontent.com/1385443/205170726-3045f823-9d05-4002-b3ac-474ce991882e.gif)

rdwebdesign commented:
Note:

I think the issue will be fixed by https://github.com/pi-hole/PADD/pull/272

yubiuser commented:
Yes. This should be addressed in https://github.com/pi-hole/PADD/pull/272
